File: dbase.js
/* dbase.js */
const sqlite = require("sqlite3");

function getDB(){
    return (new sqlite.Database(process.env.DATABASE));
}

function procData(data){ 
    var keys = [];
    var vals = [];
    var rand = [];

    for (key in data) {
        keys.push(key); vals.push(data[key]);
        rand.push("?");
    }

    return {
        keys: keys,
        vals: vals,
        str: rand
    };
}

module.exports = {
    // callback(sqlite_obj)
    insert: function (table, data) {
        return new Promise((res, err) => { 
            var obj = procData(data);

            var sql = "INSERT INTO " + table + "("+
                obj.keys.join(",") + ") VALUES ("+
                obj.str.join(",") + ")";
           
            var db = getDB();
            db.run(sql, obj.vals, () => res());
            db.close(); 
        });
    },
    
    // fields = [field1, field2]
    // where = "u_id=?"; params = [u_id]
    // callback(rows) 
    select: function (table, where, params, fields) {
        return new Promise((res, err) => {
            var vals = "*";
            if (fields) {
                vals = fields.join(",");
            }

            var sql = "SELECT "+vals+" FROM "+table+" WHERE "+where;

            var db = getDB();
            db.all(sql, params, function(db_err, rows){
                if(db_err){
                    console.log("err: dbase.select - "+db_err);
                    return err();
                }
                res(rows);
            }); 
            db.close();

        });
    },
    
    // callback(sqlite_obj)
    update: function (table, data, where, params) {
        return new Promise((res, err) => {
            var obj = procData(data);
            keys = obj.keys.map(function(key){ return key+" = ?"; } ).join(",");

            var sql = "UPDATE "+table+" SET "+keys+" WHERE "+where;

            var db = getDB();
            db.run(sql, obj.vals.concat(params), () => res());
            db.close();
        });
    },

    // callback(sqlite_obj)
    delete: function (table, where, params) {
        return new Promise((res, err) => {
            var sql = "DELETE FROM "+table+" WHERE "+where;

            var db = getDB();
            db.run(sql, params, () => res());
            db.close();
        });
    },

    setup: function(schema){
        return new Promise((res, err) => {
            var db = getDB(); 
            db.exec(schema, ()=>res());
            db.close();
        });
    }
};

File: dbase.py
"""wrapper for sqlite library"""
import sqlite3

DB = "data.db"

def get_conn():
    """get a connection to sqlite"""
    return sqlite3.connect(DB)

### return [rows]
def select(table, where="", vals=None):
    """select from sqlite db"""
    conn = get_conn()
    cur = conn.cursor()

    if not where:
        where = ""

    if not vals:
        vals = []

    cur.execute("select * from %s %s"%(table, where), vals)
    res = cur.fetchall()
    fld = cur.description
    conn.close()

    out = []
    for val in res:
        row, cnt = {}, len(val)

        for i in range(cnt):
            row[fld[i][0]] = val[i]
        out.append(row)

    return out

### return True
def insert(table, obj):
    """insert into sqlite db"""
    conn = get_conn()
    cur = conn.cursor()

    fstr, vals = [], []
    for key in obj.keys():
        fstr.append(key)
        vals.append(obj.get(key, ""))

    qry = "insert into %s(%s) values(%s)"%(
        table,
        ",".join(fstr),
        ",".join(["?"]*len(vals)))

    cur.execute(qry, vals)
    conn.commit()
    conn.close()
    return True

### return True
def delete(table, where, vals):
    """delete from sqlite db"""
    conn = get_conn()
    cur = conn.cursor()

    cur.execute("delete from %s %s"%(table, where), vals)
    conn.commit()
    conn.close()
    return True

### return True
def update(table, where, wvals, obj):
    """update vals in sqlite db"""
    conn = get_conn()
    cur = conn.cursor()

    vals, flds = [], []
    for key in obj.keys():
        flds.append(key+"=?")
        vals.append(obj[key])

    qry = "update %s set %s %s"%(
        table,
        ",".join(flds),
        where)

    cur.execute(qry, vals+wvals)
    conn.commit()
    conn.close()
    return True
